src/hooks/useNodeViews.tsx

```diff
@@ -0,0 +1,82 @@
+import React, { useCallback, useMemo, useState } from "react";
+import { createPortal } from "react-dom";
+
+import {
+  PORTAL_REGISTRY_ROOT_KEY,
+  PortalRegistry,
+  PortalRegistryContext,
+} from "../contexts/PortalRegistryContext.js";
+import {
+  ReactNodeViewConstructor,
+  RegisterElement,
+  createReactNodeViewConstructor,
+} from "../nodeViews/createReactNodeViewConstructor.js";
+
+type Props = {
+  portals: PortalRegistry;
+};
+
+function NodeViews({ portals }: Props) {
+  const rootRegisteredPortals = portals[PORTAL_REGISTRY_ROOT_KEY];
+  const rootPortals = useMemo(
+    () =>
+      rootRegisteredPortals
+        ?.sort((a, b) => a.getPos() - b.getPos())
+        .map(({ portal }) => portal),
+    [rootRegisteredPortals]
+  );
+
+  return (
+    <PortalRegistryContext.Provider value={portals}>
+      {rootPortals}
+    </PortalRegistryContext.Provider>
+  );
+}
+
+export function useNodeViews(
+  nodeViews: Record<string, ReactNodeViewConstructor>
+) {
+  const [portals, setPortals] = useState({} as PortalRegistry);
+
+  const registerPortal: RegisterElement = useCallback(
+    (registrationKey, getPos, child, container, key) => {
+      const portal = createPortal(child, container, key);
+      setPortals((oldPortals) => {
+        const oldChildPortals = oldPortals[registrationKey] ?? [];
+        const newChildPortals = oldChildPortals.concat({ getPos, portal });
+        return {
+          ...oldPortals,
+          [registrationKey]: newChildPortals,
+        };
+      });
+
+      return () => {
+        setPortals((oldPortals) => {
+          const oldChildPortals = oldPortals[registrationKey] ?? [];
+          const newChildPortals = oldChildPortals.filter(
+            ({ portal: p }) => p !== portal
+          );
+          return {
+            ...oldPortals,
+            [registrationKey]: newChildPortals,
+          };
+        });
+      };
+    },
+    []
+  );
+
+  const reactNodeViews = useMemo(() => {
+    const nodeViewEntries = Object.entries(nodeViews);
+    const reactNodeViewEntries = nodeViewEntries.map(([name, constructor]) => [
+      name,
+      createReactNodeViewConstructor(constructor, registerPortal),
+    ]);
+    return Object.fromEntries(reactNodeViewEntries);
+  }, [nodeViews, registerPortal]);
```

Comment:
I know we talked a lot outside of GitHub about how to make sure we re-register when nodes get new keys, but I can't remember if we've resolved it.

Can we require the state (editor or plugin) as an argument to this hook and use the registry seed as a dependency to force the recreation of the node views?

Reply:
As it is, it works correctly. When a node view is updated, if its key has changed, it will return false from the update method, forcing it and its children to be recreated with the new keys. This lets us incrementally refresh keys as nodes are changed, which seems better than forcing ProseMirror to rebuild all of the node views every time the editor state is reinitialized. If we wanted to instead force rebuild all of the node views, I think what you're suggesting would work, though!

Reply:
Ah, right, I forgot that we had originally discussed trying to keep track of the host ancestor's key, but eventually settled on nodes keeping track of their own key. Assuming the document changes at all, the top-level nodes have a chance to check their keys and force a redraw for themselves (and their descendants) if it changed.